### ISSUE TITLE
update datasource cause error with webgl backend 

### ISSUE THREAD

**Issue by ruoyu0088** [bokeh/bokeh#8346]
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)

windows 10, chrome, bokeh 0.13.0, python 3.7

#### Description of expected behavior and the observed behavior

update data attribute of ColumnDataSource cause javascript error with webgl backend.

#### Complete, minimal, self-contained example code that reproduces the issue

```
import numpy as np
from bokeh.plotting import figure, curdoc
from bokeh.models import ColumnDataSource
from bokeh.models.widgets import Button
from bokeh.layouts import row, column, gridplot

doc = curdoc()

change_button = Button(label="change")
source = ColumnDataSource(data={"x":[1, 2, 3], "y":[4, 5, 6], "x1":[1, 2, 3], "y1":[1, 2, 1]})
backend = curdoc().session_context.request.arguments.get("backend", [b"canvas"])[0].decode()
print(backend)

def create_figure():
    fig1 = figure(plot_height=100, plot_width=200, output_backend=backend)
    fig1.line(x="x", y="y", source=source)
    fig2 = figure(plot_height=100, plot_width=200, output_backend=backend)
    fig2.line(x="x1", y="y1", source=source)
    grid = gridplot([[fig1], [fig2]])
    return grid

def change_data():
    x = np.linspace(-1, 1)
    y = x**2
    source.data = dict(x=x, y=y, x1=x*2, y1=y*2)


figure_box = column(create_figure())
layout = column(change_button, figure_box)
change_button.on_click(change_data)
doc.add_root(layout)
```

#### Screenshots or screencasts of the bug in action

run the serve by `bokeh serve change_data.py`, and if I choose canvas backend, and click the change button:

http://localhost:5006/change_data?backend=canvas

![image](https://user-images.githubusercontent.com/826871/47197070-d8c6c980-d39e-11e8-9134-c3f39512d51d.png)

if I choose webgl backend, and click the change button:

http://localhost:5006/change_data?backend=webgl

![image](https://user-images.githubusercontent.com/826871/47197073-debcaa80-d39e-11e8-8f4e-61971a455650.png)

and here is the output console of Chrome:

```
logging.js:130 [bokeh] setting log level to: 'info'
connection.js:183 [bokeh] Websocket connection 0 is now open
widget_box.js:98 [bokeh] WidgetBox mode is fixed, but no width specified. Using default of 300.
document.js:182 [bokeh] document idle at 473 ms
index.js:48 Bokeh items were rendered successfully
gloo2.js:227 Uncaught Error: RuntimeError:OpenGL got errors (after draw): 1282
    at l (gloo2.js:227)
    at i.draw (gloo2.js:649)
    at e.draw (line.js:168)
    at e.t.render (base.js:52)
    at e.t.render (glyph.js:60)
    at t.render (glyph_renderer.js:234)
    at t._paint_levels (plot_canvas.js:770)
    at t.paint (plot_canvas.js:734)
    at t.repaint (plot_canvas.js:666)
    at t.<anonymous> (plot_canvas.js:547)
l @ gloo2.js:227
i.draw @ gloo2.js:649
e.draw @ line.js:168
t.render @ base.js:52
t.render @ glyph.js:60
t.render @ glyph_renderer.js:234
t._paint_levels @ plot_canvas.js:770
t.paint @ plot_canvas.js:734
t.repaint @ plot_canvas.js:666
(anonymous) @ plot_canvas.js:547
t.emit @ signaling.js:55
e.emit @ signaling.js:68
(anonymous) @ plot_canvas.js:130
s @ throttle.js:29
requestAnimationFrame (async)
(anonymous) @ throttle.js:38
t.request_paint @ plot_canvas.js:91
t.request_render @ plot_canvas.js:87
t.request_render @ renderer.js:27
t.set_data @ glyph_renderer.js:138
(anonymous) @ glyph_renderer.js:85
t.emit @ signaling.js:55
e.emit @ signaling.js:68
u._setv @ has_props.js:216
u.setv @ has_props.js:239
set @ has_props.js:91
t.compute_indices @ cds_view.js:62
(anonymous) @ cds_view.js:37
t.emit @ signaling.js:55
e.emit @ signaling.js:68
u._setv @ has_props.js:216
u.setv @ has_props.js:239
P.apply_json_patch @ document.js:752
t._handle_patch @ session.js:70
t.handle @ session.js:19
t._steady_state_handler @ connection.js:262
ACK.t.msgtype._current_handler @ connection.js:242
t._on_message @ connection.js:204
i.socket.onmessage @ connection.js:61
change_data?backend=webgl:1 [.WebGL-0000024C4B523B80]GL ERROR :GL_INVALID_OPERATION : glDrawElements: attempt to access out of range vertices in attribute 2
gloo2.js:227 Uncaught Error: RuntimeError:OpenGL got errors (after draw): 1282
    at l (gloo2.js:227)
    at i.draw (gloo2.js:649)
    at e.draw (line.js:168)
    at e.t.render (base.js:52)
    at e.t.render (glyph.js:60)
    at t.render (glyph_renderer.js:234)
    at t._paint_levels (plot_canvas.js:770)
    at t.paint (plot_canvas.js:734)
    at t.repaint (plot_canvas.js:666)
    at t.<anonymous> (plot_canvas.js:547)
l @ gloo2.js:227
i.draw @ gloo2.js:649
e.draw @ line.js:168
t.render @ base.js:52
t.render @ glyph.js:60
t.render @ glyph_renderer.js:234
t._paint_levels @ plot_canvas.js:770
t.paint @ plot_canvas.js:734
t.repaint @ plot_canvas.js:666
(anonymous) @ plot_canvas.js:547
t.emit @ signaling.js:55
e.emit @ signaling.js:68
(anonymous) @ plot_canvas.js:130
s @ throttle.js:29
requestAnimationFrame (async)
(anonymous) @ throttle.js:41
setTimeout (async)
(anonymous) @ throttle.js:41
t.request_paint @ plot_canvas.js:91
t.request_render @ plot_canvas.js:87
(anonymous) @ plot_canvas.js:551
t.emit @ signaling.js:55
e.emit @ signaling.js:68
u._setv @ has_props.js:216
u.setv @ has_props.js:239
t.update @ data_range1d.js:230
t.update_dataranges @ plot_canvas.js:244
t.paint @ plot_canvas.js:695
t.repaint @ plot_canvas.js:666
(anonymous) @ plot_canvas.js:547
t.emit @ signaling.js:55
e.emit @ signaling.js:68
(anonymous) @ plot_canvas.js:130
s @ throttle.js:29
requestAnimationFrame (async)
(anonymous) @ throttle.js:38
t.request_paint @ plot_canvas.js:91
t.request_render @ plot_canvas.js:87
t.request_render @ renderer.js:27
t.set_data @ glyph_renderer.js:138
(anonymous) @ glyph_renderer.js:85
t.emit @ signaling.js:55
e.emit @ signaling.js:68
u._setv @ has_props.js:216
u.setv @ has_props.js:239
set @ has_props.js:91
t.compute_indices @ cds_view.js:62
(anonymous) @ cds_view.js:37
t.emit @ signaling.js:55
e.emit @ signaling.js:68
u._setv @ has_props.js:216
u.setv @ has_props.js:239
P.apply_json_patch @ document.js:752
t._handle_patch @ session.js:70
t.handle @ session.js:19
t._steady_state_handler @ connection.js:262
ACK.t.msgtype._current_handler @ connection.js:242
t._on_message @ connection.js:204
i.socket.onmessage @ connection.js:61
change_data?backend=webgl:1 [.WebGL-0000024C4B523B80]GL ERROR :GL_INVALID_OPERATION : glDrawElements: attempt to access out of range vertices in attribute 2
```

**Comment by mattpap** [comment]
My guess is that GL glyphs, that mirror canvas glyphs when webgl backend is selected, aren't updated (or updated properly) when data changes. If that's the case, it would make this issue pretty easy to fix.

**Comment by bryevdv** [comment]
@ruoyu0088 why did you close the issue?

**Comment by mattpap** [comment]
It's fixed (most likely in 2.4), but still a regression test would be in order.
